.github/workflows/build-containers.yml

```diff
@@ -0,0 +1,83 @@
+  release:
+    types: [published]
+  # And let's build but not release Spack containers on a PR to develop.
+  pull_request:
+    branches:
+      - develop
+
+jobs:
+  deploy-images:
+    runs-on: ubuntu-latest 
+    strategy:
+      # Even if one container fails to build we still want the others
+      # to continue their builds.
+      fail-fast: false
+      # A matrix of Dockerfile paths, associated tags, and which architectures
+      # they support.
+      matrix:
+        dockerfile: [[amazon-linux, amazonlinux-2.dockerfile, 'linux/amd64,linux/arm64'], 
+                     [centos7, centos-7.dockerfile, 'linux/amd64,linux/arm64'],
+                     [leap15, leap-15.dockerfile, 'linux/amd64,linux/arm64'],
+                     [ubuntu-xenial, ubuntu-1604.dockerfile, 'linux/amd64,linux/arm64'],
+                     [ubuntu-bionic, ubuntu-1804.dockerfile, 'linux/amd64,linux/arm64']]
+    name: Build ${{ matrix.dockerfile[0] }}
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Set Container Tag Normal (PR & Nightly)
+        run: |
+          container="ghcr.io/spack/${{ matrix.dockerfile[0]}}:latest"
+          echo "container=${container}" >> $GITHUB_ENV
+          echo "versioned=${container}" >> $GITHUB_ENV
+
+        # On a new release create a container with the same tag as the release.
+      - name: Set Container Tag on Release
+        if: github.event_name == 'release'
+        run: |
+          versioned="$ghcr.io/spack/${{matrix.dockerfile[0]}}:${GITHUB_REF##*/}"
+          echo "versioned=${versioned}" >> $GITHUB_ENV
+
+      - name: Check ${{ matrix.dockerfile[1] }} Exists
+        run: |
+          printf "Preparing to build ${{ env.container }} from ${{ matrix.dockerfile[1] }}"
+          if [ ! -f "share/spack/docker/${{ matrix.dockerfile[1]}}" ]; then
+              printf "Dockerfile ${{ matrix.dockerfile[0]}} does not exist"
+              exit 1;
+          fi
+
+      - name: Log in to GitHub Container Registry
+        # Don't log into registry on pull request.
+        if: github.event_name != 'pull_request'
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: spack
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Build ${{ matrix.dockerfile[1] }}
+        uses: docker/build-push-action@v2
+        with:
+          file: share/spack/docker/${{matrix.dockerfile[1]}}
+          platforms: ${{ matrix.dockerfile[2] }}
+          push: false
+          tags: |
+            ${{ env.container }}
+            ${{ env.versioned }}
+
+      - name: Deploy ${{ matrix.dockerfile[1] }}
+        # And don't push the container on a pull request.
+        if: github.event_name != 'pull_request'
+        run: docker push ghcr.io/spack/${{ matrix.dockerfile[0]}} --all-tags
```

Comment:
I think we'll need a token stored as a secret to be able to do this.

Reply:
Also, just as a note (absolutely not a request on this PR), when #21910 is merged we could drop the explicit Docker recipes and let `spack containerize` generate them during CI + upload the artifacts. This will reduce the places we need to check to update Dockerfiles to just the `containerize` Jinja 2 templates.

Reply:
> (At the moment x86_64 and Arm64 because ppc64le is throwing an error within archspec.)

I'll check that as soon as I can. For what is worth I was able to use Spack with `docker buildx` in [another repository](https://github.com/alalazo/spack-bootstrap-mirrors) to build recipes - so wondering what would be different in this case.

Reply:
> I think we'll need a token stored as a secret to be able to do this.

On line 55 is where we log into the GitHub Container Registry which is actually able to just use the GitHub token assigned to the workflow. We will absolutely have to add additional steps and secrets if we want to also push to Dockerhub, etc...

Reply:
> > (At the moment x86_64 and Arm64 because ppc64le is throwing an error within archspec.)
> 
> I'll check that as soon as I can. For what is worth I was able to use Spack with `docker buildx` in [another repository](https://github.com/alalazo/spack-bootstrap-mirrors) to build recipes - so wondering what would be different in this case.

Interesting! I see the patch you mentioned in the Slack conversation so I'll try that out and see if that fixes it!

Reply:
> On line 55 is where we log into the GitHub Container Registry which is actually able to just use the GitHub token assigned to the workflow.

OH, interesting! I must have missed this blog post: https://github.blog/changelog/2021-03-24-packages-container-registry-now-supports-github_token/

Reply:
Oh yeah it’s super cool that you can now just use the GitHub token! It makes it really nice for writing build templates/teaching people to use actions with containers since they no longer need a PAT!